toolkit/tools/internal/pkgjson/pkgjson.go

```diff
@@ -339,78 +338,56 @@ func (interval *PackageVerInterval) Compare(other *PackageVerInterval) (result i
-	var (
-		queryUpperValid, queryLowerValid, superset bool
-	)
-	if interval.LowerBound.Compare(queryInterval.UpperBound) > 0 || interval.UpperBound.Compare(queryInterval.LowerBound) < 0 {
+func (interval *PackageVerInterval) intersects(queryInterval *PackageVerInterval) (intersects bool) {
+	if interval.LowerBound.GT(queryInterval.UpperBound) || interval.UpperBound.LT(queryInterval.LowerBound) {
 		// No overlap at all
 		return false
 	}
 
-	// Check lower bound of the query
-	switch {
-	case queryInterval.LowerBound.Compare(interval.UpperBound) == 0:
-		// Check if the query lower bound touches the interval upper bound exactly
-		if interval.UpperInclusive && queryInterval.LowerInclusive {
-			queryLowerValid = true
-		}
-	case queryInterval.LowerBound.Compare(interval.LowerBound) == 0:
-		// Check if both lower bounds are the same
-		if interval.LowerInclusive || !queryInterval.LowerInclusive {
-			queryLowerValid = true
-		}
-	default:
-		// Otherwise simple check
-		queryLowerValid = interval.versionInInterval(queryInterval.LowerBound)
+	// Handle equal versions but open intervals
+	if interval.LowerBound.EQ(queryInterval.UpperBound) && (!interval.LowerInclusive || !queryInterval.UpperInclusive) {
+		return false
 	}
 
-	// Check the upper bound of the query
-	switch {
-	case queryInterval.UpperBound.Compare(interval.LowerBound) == 0:
-		// Check if the query upper bound touches the interval lower bound exactly
-		if interval.LowerInclusive && queryInterval.UpperInclusive {
-			queryUpperValid = true
-		}
-	case queryInterval.UpperBound.Compare(interval.UpperBound) == 0:
-		// Check if both upper bounds are the same
-		if interval.UpperInclusive || !queryInterval.UpperInclusive {
-			queryLowerValid = true
-		}
-	default:
-		// Otherwise simple check
-		queryUpperValid = interval.versionInInterval(queryInterval.UpperBound)
+	if interval.UpperBound.EQ(queryInterval.LowerBound) && (!interval.UpperInclusive || !queryInterval.LowerInclusive) {
+		return false
 	}
 
-	// Check if the query interval is a superset of the provided interval
-	superset = queryInterval.LowerBound.Compare(interval.LowerBound) < 0 && queryInterval.UpperBound.Compare(interval.UpperBound) > 0
+	return true
+}
+
+// GuaranteedSatisfies returns true if every version in the interval is guaranteed to satisfy the query. A pair of
+// intervals may overlap, but not guarantee that the query interval is satisfied.
+// i.e., interval: [1, 3], query: [2, 4] will overlap, but the valid version '1' does not satisfy the query.
+func (interval *PackageVerInterval) GuaranteedSatisfies(queryInterval *PackageVerInterval) (valid bool) {
+	return queryInterval.contains(interval)
+}
 
-	return queryUpperValid || queryLowerValid || superset
+// Satisfies returns true if there exists some version in the interval that satisfies the query interval.
+func (interval *PackageVerInterval) Satisfies(queryInterval *PackageVerInterval) (valid bool) {
```

Comment:
The `Satisfies` vs `GuaranteedSatisfies` naming feel a bit confusing. The mathematical terms you're using in the private functions seem more descriptive. What do you think about using `Contains` and `Intersects`/`Overlaps` directly? In the former case I guess we'd need to reverse the order of how we call `GuaranteedSatisfies` now but it feel like a small price to pay for clarity. Especially since this piece of code needs extra precision to avoid making a mistake.